capture.js

```diff
@@ -0,0 +1,53 @@
+const capture = {
+
+  init() {
+    this.addListeners();
+    this.prevSite = null;
+    this.nextSite = null;
+  },
+
+  addListeners() {
+    // listen for each HTTP response
+    browser.webRequest.onResponseStarted.addListener(this.collectRequests, {urls: ['<all_urls>']}
+    );
+    // listen for when user visits a new site
+    browser.tabs.onUpdated.addListener(this.checkNewSite);
+  },
+
+  async collectRequests(response) {
+    const tab = await browser.tabs.get(response.tabId);
+    let tabUrl = new URL(tab.url);
+    tabUrl = tabUrl.hostname;
+    let targetUrl = new URL(response.url);
+    targetUrl = targetUrl.hostname;
+    let originUrl = new URL(response.originUrl);
+    originUrl = originUrl.hostname;
+
+    // capture third party requests
+    if (targetUrl !== originUrl) {
+      const thirdPartyData = {
+        document: tabUrl,
+        target: targetUrl,
+        origin: originUrl,
+        requestTime: response.timeStamp
+      };
+      console.log('storage.thirdPartyRequest:', tabUrl, thirdPartyData);
+    }
+  },
+
+  checkNewSite(tabId, changeInfo, tab) {
+    if (changeInfo.url) {
+      this.nextSite = changeInfo.url;
+    }
+    //favicon not available until tab status is complete; opening a new tab (about:blank) does not trigger a call to storage API
+    if ((this.nextSite !== this.prevSite) && (changeInfo.status === 'complete') && (this.nextSite !== null) && (this.nextSite !== 'about:blank')) {
```

Comment:
I'm not 100% certain that this API will provide a url always as it seems to be about state tracking of tab updates. (The MDN docs aren't actually clear or leave it to ambiguity which suggest anything optional could be so).

Because all the [changeInfo](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/tabs/onUpdated#changeInfo) is optional I would only be checking for changes when completed:

```js
async collectRequests(tabId, updated) {
  const tab = await = browser.tabs.get(tabId);
  const tabUrl = new URL(tab.url);
  if (updated.status === "completed") {
    console.log('storage.firstPartyUpdate:', tabUrl.hostname, tab);
  }
})
```

My understanding is the API will only provide urls when they have changed which means the storage should take care of piecemeal building up what a tab represents.